src/mongo_odm/nvp.hpp

```diff
@@ -56,16 +74,83 @@
     MONGO_ODM_MAKE_KEYS(Base, MONGO_ODM_NVP(_id), __VA_ARGS__)
 
 #define MONGO_ODM_KEY(value) mongo_odm::hasCallIfFieldIsPresent<decltype(&value), &value>::call()
+#define MONGO_ODM_KEY_BY_VALUE(value) \
+    mongo_odm::hasCallIfFieldIsPresent<decltype(value), value>::call()
+
+#define MONGO_ODM_CHILD1(base, field1) MONGO_ODM_KEY_BY_VALUE(&base::field1)
+
+#define MONGO_ODM_CHILD2(base, field1, field2)                                               \
+    makeNvpWithParent(                                                                       \
+        MONGO_ODM_KEY_BY_VALUE(                                                              \
+            &std::decay_t<typename decltype(MONGO_ODM_CHILD1(base, field1))::type>::field2), \
+        MONGO_ODM_CHILD1(base, field1))
+
+#define MONGO_ODM_CHILD3(base, field1, field2, field3)                                         \
+    makeNvpWithParent(MONGO_ODM_KEY_BY_VALUE(&std::decay_t<typename decltype(MONGO_ODM_CHILD2( \
+                                                 base, field1, field2))::type>::field3),       \
+                      MONGO_ODM_CHILD2(base, field1, field2))
+
+#define PASTE_IMPL(s1, s2) s1##s2
+#define PASTE(s1, s2) PASTE_IMPL(s1, s2)
+
+#define MONGO_ODM_CHILD(type, ...) PASTE(MONGO_ODM_CHILD, PP_NARG(__VA_ARGS__))(type, __VA_ARGS__)
 
 namespace mongo_odm {
```

Comment:
I think it's going to be a good idea to cut out an impl namespace inside of mongo_odm and move all of the utility stuff we have kicking around into that.  Consider moving any types/functions it would never make sense for a user to directly name into that namespace.